modules/drivers/sensors/hx711/modHX711c.c

```diff
@@ -0,0 +1,76 @@
+// Copyright © 2023 by Thorsten von Eicken.
+
+#include "xsPlatform.h"
+#include "xsmc.h"
+#include "modGPIO.h"
+#include "mc.xs.h" // for xsID_* constants
+
+#define xsmcVar(x) xsVar(x)
+
+typedef struct {
+  modGPIOConfigurationRecord clk;
+  modGPIOConfigurationRecord din;
+  int gain;
+} hx711_data;
+
+void xs_HX711_init(xsMachine *the) {
+  if (xsmcArgc != 3) xsUnknownError("invalid arguments");
+  int clk_pin = xsmcToInteger(xsArg(0));
+  int din_pin = xsmcToInteger(xsArg(1));
+
+  hx711_data *data = c_malloc(sizeof(hx711_data));
+  if (data == NULL) xsUnknownError("can't allocate data");
+  data->gain = xsmcToInteger(xsArg(2));
+  if (modGPIOInit(&data->clk, NULL, clk_pin, kModGPIOOutput))
+    xsUnknownError("can't init clk pin");
+  modGPIOWrite(&data->clk, 0);
+  if (modGPIOInit(&data->din, NULL, din_pin, kModGPIOInput))
+    xsUnknownError("can't init dat pin");
+
+  xsmcSetHostData(xsThis, data);
+}
+
+void xs_HX711_destructor(void *hostData) {
+  hx711_data *data = hostData;
```

Comment:
The destructor can be called with `hostData` set to `NULL`. This happens when the virtual machine is deleted. It is rare. ;) It can also happen if an exception occurs in the constructor before the host data is set. So, the best practice is to always check for `NULL`